docs: adding rehydration doc and updates to building apps overview (gatsby for apps workflow) (#19214)
gatsbyjs/gatsby: docs/docs/react-rehydration.md

---
title: Understanding React Rehydration
---

> One of the central ideas of Gatsby is that HTML content is statically generated--using React DOM server-side APIs. A follow up key feature is that this static HTML content can then be _enhanced_ with client-side JavaScript via React hydration which allows for [app-like features](/docs/adding-app-and-website-functionality/) in Gatsby sites.

Reviewer: ```suggestion
One of the central ideas of Gatsby is that HTML content is statically generated using React DOM server-side APIs. A follow up key feature is that this static HTML content can then be _enhanced_ with client-side JavaScript via React hydration which allows for [app-like features](/docs/adding-app-and-website-functionality/) in Gatsby sites.
```

The steps taken to deliver a built site to the browser are discussed below:

## Build and render static assets

Running `gatsby build` starts up a Node.js server that processes your site: creating a GraphQL schema for your site, fetching data that your pages will pull in by [extracting queries](/docs/query-extraction/) from your code and [executing them](/docs/query-execution/), and [rendering each page's HTML](/docs/html-generation/).

_The ["Overview of the Gatsby Build Process" conceptual guide](/docs/overview-of-the-gatsby-build-process/) helps explain what's happening at each step in the build process_

> All of this data is gathered during the build and written into the `/public` folder. You can [customize Gatsby's configurations](/docs/customization/) for `babel`, `webpack`, as well as the HTML generated by your build in order for finer grained tweaks in regards to how your site gets built.

Reviewer 2: ```suggestion
All of this data is gathered during the build and written into the `/public` folder. You can [customize Gatsby's configurations](/docs/customization/) for `babel` and `webpack`, as well as the HTML generated by your build, in order to tweak how your site gets built.
```

Reviewer: Should we not code format Babel and webpack? And capital letter on Babel?

```suggestion
All of this data is gathered during the build and written into the `/public` folder. You can [customize Gatsby's configurations](/docs/customization/) for Babel, webpack, as well as the HTML generated by your build in order for finer grained tweaks in regards to how your site gets built.
```

> ## Invoke the `ReactDOM.hydrate` method

Reviewer: I miss here the info: is this done by Gatsby or should I need to place this command in my files

The `hydrate()` method is called from `ReactDOM`, which according to the [React docs](https://reactjs.org/docs/react-dom.html#hydrate) is:

> Same as [render()](https://reactjs.org/docs/react-dom.html#render), but is used to hydrate a container whose HTML contents were rendered by [ReactDOMServer](https://reactjs.org/docs/react-dom-server.html).

> This means that the browser can "pick up" where the server left off with the contents created by Gatsby in the `/public` folder and mount the site in the browser just like any other React app would be. Since the data and structure of the pages is already written out, it's not necessary for Gatsby to go to another asking for HTML or other data.

Reviewer 2: What do we mean by `another` here? Might be a noun missing.

---
title: Understanding React Rehydration
---

> One of the central ideas of Gatsby is that HTML content is statically generated using React DOM server-side APIs. A follow up key feature is that this static HTML content can then be _enhanced_ with client-side JavaScript via React hydration which allows for [app-like features](/docs/adding-app-and-website-functionality/) in Gatsby sites.

Reviewer 3: ```suggestion
One of the central ideas of Gatsby is that HTML content is statically generated using React DOM server-side APIs. Another key feature is that this static HTML content can then be _enhanced_ with client-side JavaScript via React hydration, which allows for [app-like features](/docs/adding-app-and-website-functionality/) in Gatsby sites.
```

The steps taken to deliver a built site to the browser are discussed below:

## Build and render static assets

> Running `gatsby build` starts up a Node.js server that processes your site: creating a GraphQL schema for your site, fetching data that your pages will pull in by [extracting queries](/docs/query-extraction/) from your code and [executing them](/docs/query-execution/), and [rendering each page's HTML](/docs/html-generation/).

Reviewer 3: ```suggestion
Running `gatsby build` starts up a Node.js server that processes your site: it creates a GraphQL schema, fetches data that your pages will pull in by [extracting queries](/docs/query-extraction/) from your code and [executing them](/docs/query-execution/), and it then [renders each page's HTML](/docs/html-generation/).
```

> _The ["Overview of the Gatsby Build Process" conceptual guide](/docs/overview-of-the-gatsby-build-process/) helps explain what's happening at each step in the build process_

Reviewer 3: ```suggestion
_The ["Overview of the Gatsby Build Process" conceptual guide](/docs/overview-of-the-gatsby-build-process/) helps explain what's happening at each step in the build process._
```

All of this data is gathered during the build and written into the `/public` folder. You can [customize Gatsby's configurations](/docs/customization/) for Babel and webpack, as well as the HTML generated by your build, in order to tweak how your site gets built.

> ## Invoke the `ReactDOM.hydrate` method

Reviewer 3: ```suggestion
## The `ReactDOM.hydrate` method
```
I'd remove the word "invoke" from the heading as it does sort-of set up a reader expectation that they'd be invoking it.

The `hydrate()` method is called internally by Gatsby from `ReactDOM`, which according to the [React docs](https://reactjs.org/docs/react-dom.html#hydrate) is:

> Same as [render()](https://reactjs.org/docs/react-dom.html#render), but is used to hydrate a container whose HTML contents were rendered by [ReactDOMServer](https://reactjs.org/docs/react-dom-server.html).

_**Note**: if you need to, the hydrate method can be replaced with a custom function by using the [`replaceHydrationFunction` Browser API](/docs/browser-apis/#replaceHydrateFunction)._

> This means that the browser can "pick up" where the server left off with the contents created by Gatsby in the `/public` folder and mount the site in the browser just like any other React app would be. Since the data and structure of the pages is already written out, it's not necessary for Gatsby to go to another server asking for HTML or other data.

Reviewer 3: ```suggestion
This means that the browser can "pick up" where the server left off with the contents created by Gatsby in the `/public` folder and render the site in the browser just like any other React app would. Since the data and structure of the pages is already written out, it's not necessary for Gatsby to go to another server asking for HTML or other data.
```

## Transfer rendering to the React reconciler

> After ReactDOM hydrates the content, the React reconciler can take over and diff the tree of elements. It then becomes responsible for making updates in the UI based on changing state or props.

Reviewer 3: ```suggestion
After ReactDOM hydrates the content, the [React reconciler](https://reactjs.org/docs/reconciliation.html) can take over and [diff](https://en.m.wikipedia.org/wiki/Diff) the tree of elements. It then becomes responsible for making updates in the UI based on changing state or props.
```
